### PR TITLE
Decompile-func_us_80171864

### PR DESCRIPTION
**Pull request by Onenutmcgee**
Decompilation of func_us_80171864
Adding ET_Ghost for the Entity.ext union.

Thanks to @sozud for setting up a PSP environment for me.
Thanks to @joshlory for figuring out the last bit.

### CODE REVIEW

File: src/servant/tt_001/F84.c [hunk 1]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+                    self, *(&D_us_8017050C + ((D_us_80173810.level / 10) * 3)),
+                    1, &D_us_80173810);
+                self->hitboxWidth = 8;
+                self->hitboxHeight = 8;
+            }
+        } else {
+            self->ext.ghost.unk86 = 0;
+            if (self->ext.ghost.unk92->entityId == 0xDA) {
+                self->ext.ghost.unk92->params = 1;
+                D_us_80170028(self->ext.ghost.unk92);
```
Comment (sozud) [inline src/servant/tt_001/F84.c:236]: Seems like this should be g_ServantDesc.XXXXX()

File: src/servant/tt_001/F84.c [hunk 2]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+            }
+        }
+        self->posX.val += self->velocityX;
+        self->posY.val += self->velocityY;
+        break;
+    case 4:
+        if (!(g_Player.unk0C & 0x01000001)) {
+            if (self->ext.ghost.unk96->entityId == 0xDB) {
+                self->ext.ghost.unk96->params = 1;
+                D_us_8017002C(self->ext.ghost.unk96);
```
Comment (sozud) [inline src/servant/tt_001/F84.c:246]: g_ServantDesc.

File: src/servant/tt_001/F84.c [hunk 3]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+        self->drawMode = DRAW_TPAGE2 | DRAW_TPAGE;
+        self->drawFlags = FLAG_DRAW_UNK8;
+        SetEntityAnimation(self, &D_us_801704A8);
+        self->ext.ghost.unk84 = 0x200;
+        self->ext.ghost.unk88 = 0x80;
+        self->ext.ghost.unk8A = -1;
+        self->step++;
+        break;
+    case 1:
+        if (g_Player.unk0C & 0x01000001) {
```
Comment (sozud) [inline src/servant/tt_001/F84.c:135]: enum https://github.com/Xeeynamo/sotn-decomp/blob/fb8e5d795083256425400782835e6e7de8ef1442/include/game.h#L291

File: src/servant/tt_001/F84.c [hunk 4]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+            self->ext.ghost.unkA2 = func_us_80171284(self);
+            if (self->ext.ghost.unkA2) {
+                self->step++;
+                break;
+            }
+        }
+        break;
+    case 2:
+    case 3:
+        if (g_Player.unk0C & 0x01000001) {
```
Comment (sozud) [inline src/servant/tt_001/F84.c:191]: enum

File: src/servant/tt_001/F84.c [hunk 5]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+        D_us_801737CC =
+            func_us_80170F84(self, (s16)D_us_801737C4, D_us_801737C8);
+        if (self->step == 2) {
+            if (D_us_801737CC < 8) {
+                self->ext.ghost.unk86 = 0;
+                self->step += 1;
+            }
+        } else if (D_us_801737CC < 8) {
+            self->ext.ghost.unk86++;
+            temp_v1 = *(&D_us_80170508 + ((D_us_80173810.level / 10) * 3));
```
Comment (sozud) [inline src/servant/tt_001/F84.c:221]: Should be an array access

`D_us_80170508[((D_us_80173810.level / 10) * 3)]` maybe ?

File: src/servant/tt_001/F84.c [hunk 6]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+            if (self->ext.ghost.unk86 == (temp_v1 - 0x1E)) {
+                self->ext.ghost.unk92 = func_us_80171568(self, 0);
+            } else if (temp_v1 < self->ext.ghost.unk86) {
+                self->ext.ghost.unk86 = 0;
+                g_api.func_8011A3AC(
+                    self, *(&D_us_8017050C + ((D_us_80173810.level / 10) * 3)),
```
Comment (sozud) [inline src/servant/tt_001/F84.c:227]: another array access

File: src/servant/tt_001/F84.c [hunk 7]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+            } else if (self->ext.ghost.unk86 >= 0x3D) {
+                self->ext.ghost.unk86 = 0;
+                self->step++;
+            }
+            break;
+        }
+        break;
+    case 5:
+        if (g_Player.unk0C & 0x01000001) {
+            if (self->velocityY > (s32)0xFFFE8000) {
```
Comment (sozud) [inline src/servant/tt_001/F84.c:295]: FIX

File: include/entity.h [hunk 8]
```diff
@@ -671,6 +671,23 @@ typedef struct {
...
+    /* 0x7E */ s16 unk7E;
+    /* 0x80 */ s16 unk80;
+    /* 0x82 */ s16 unk82;
+    /* 0x84 */ s16 unk84;
+    /* 0x86 */ s16 unk86;
+    /* 0x88 */ s16 unk88;
+    /* 0x8A */ s16 unk8A;
+    /* 0x8C */ s16 unk8C;
+    /* 0x8E */ s32 pad8E;
+    /* 0x92 */ struct Entity* unk92;
```
Comment (sozud) [inline include/entity.h:685]: Not critical right now but these pointers are going to be out of alignment if compiled in 64-bit mode

File: src/servant/tt_001/F84.c [hunk 9]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+        if (!(g_Player.unk0C & 0x01000001)) {
+            self->step = 1;
+        } else {
+            self->posX.val = FIX(128.0);
+            self->posY.val = FIX(-16.0);
+        }
+        break;
+    }
+    self->ext.ghost.unk88 += self->ext.ghost.unk8A;
+    if ((s16)(self->ext.ghost.unk88 - 0x21) >= 0x5Fu) {
```
Comment (sozud) [inline src/servant/tt_001/F84.c:319]: Something like
```
    if(!(arg0->ext.ghost.unk88 >= 0x21 && arg0->ext.ghost.unk88 < 0x5F + 0x21))
```
haven't verified on psp

File: src/servant/tt_001/F84.c [hunk 10]
```diff
@@ -57,7 +77,254 @@ Entity* func_us_80171568(Entity* self, s32 entityId) {
...
+                self->posX.val = FIX(128.0);
+                self->posY.val = FIX(-16.0);
+                self->step++;
+            }
+        } else {
+            self->step = 1;
+        }
+        break;
+    case 6:
+        if (!(g_Player.unk0C & 0x01000001)) {
```
Comment (sozud) [inline src/servant/tt_001/F84.c:310]: enum

File: include/entity.h (hunk 8 above)
Comment (Onenutmcgee) [reply include/entity.h:685]: I'd like to assume by that point, we probably won't need to have these named as such (or the comment with where they are in the struct).  I can remove the comments if you'd like, but I don't really know what else to name the variables.

**Comment by Onenutmcgee** [comment]
I'll resolve these tomorrow or Monday depending on when I have time.

File: include/entity.h (hunk 8 above)
Comment (sozud) [reply include/entity.h:685]: I don't mean the comments, I'm not sure exactly what the issue will be but there's most likely going to be portability bugs when the time comes. Here's examples of static asserts we use to check alignment on structs https://github.com/Xeeynamo/sotn-decomp/blob/fb8e5d795083256425400782835e6e7de8ef1442/include/entity.h#L1958

You don't need to change this to get it merged, I just thought it should be pointed out

Comment (Onenutmcgee) [reply include/entity.h:685]: Ok.  I'll look into setting that up for this in another PR (as well as setting that up for ET_Bat as well).

File: src/servant/tt_001/F84.c (hunk 9 above)
Comment (Onenutmcgee) [reply src/servant/tt_001/F84.c]: Yeah, that change cleans up a lot on the PSP version as well.

**Comment by Onenutmcgee** [comment]
I seem to have broken this. I’ll have to look into why when I have some more time.

**Comment by Onenutmcgee** [comment]
All greens.  Should be ready for review now.

File: src/servant/tt_001/F84.c (hunk 1 above)
Comment (Onenutmcgee) [reply src/servant/tt_001/F84.c]: In order to get this one to work I had to merge in my ServantDesc data decomp change.

**Review by sozud — APPROVED** [review]
Looks good! Thanks for the revisions

**Pull request merged**